reconcile.go

```diff
@@ -234,54 +243,107 @@ func (a *Autopilot) pruneDeadServers() error {
 
 	state := a.GetState()
 
-	failed, voters, err := a.getFailedServers()
+	failed, vr, err := a.getFailedServers()
 	if err != nil || failed == nil {
 		return err
 	}
 
 	failed = a.promoter.FilterFailedServerRemovals(conf, state, failed)
 
-	// remove failed non voting servers
-	for _, srv := range failed.FailedNonVoters {
-		a.logger.Info("Attempting removal of failed server node", "id", srv.ID, "name", srv.Name, "address", srv.Address)
-		a.delegate.RemoveFailedServer(srv)
+	// remove stale non-voters
+	staleNonVoters := vr.FilterStale(failed.StaleNonVoters)
+	toRemove := a.adjudicateRemoval(staleNonVoters, vr.PotentialVoters)
+	if err = a.removeStaleServers(toRemove); err != nil {
+		return err
 	}
+	vr.Remove(toRemove)
 
-	// remove stale non voters
-	for _, id := range failed.StaleNonVoters {
-		a.logger.Debug("removing stale raft server from configuration", "id", id)
-		if err := a.removeServer(id); err != nil {
-			return err
-		}
+	// Remove stale voters
+	staleVoters := vr.FilterStale(failed.StaleVoters)
+	toRemove = a.adjudicateRemoval(staleVoters, vr.PotentialVoters)
+	if err = a.removeStaleServers(toRemove); err != nil {
+		return err
 	}
+	vr.Remove(toRemove)
 
-	maxRemoval := (voters - 1) / 2
+	// remove failed non-voters
+	failedNonVoters := vr.FilterFailed(failed.FailedNonVoters)
+	toRemove = a.adjudicateRemoval(failedNonVoters, vr.PotentialVoters)
+	a.removeFailedServers(failed.Get(toRemove, false))
+	vr.Remove(toRemove)
 
-	for _, id := range failed.StaleVoters {
-		if voters-1 < int(conf.MinQuorum) {
-			a.logger.Debug("will not remove server as it would leave less voters than the minimum number allowed", "id", id, "min", conf.MinQuorum)
-		} else if maxRemoval < 1 {
-			a.logger.Debug("will not remove server as removal of a majority of servers is not safe", "id", id)
-		} else if err := a.removeServer(id); err != nil {
-			return err
-		} else {
+	// remove failed voters
+	failedVoters := vr.FilterFailed(failed.FailedVoters)
+	toRemove = a.adjudicateRemoval(failedVoters, vr.PotentialVoters)
+	a.removeFailedServers(failed.Get(toRemove, true))
+	vr.Remove(toRemove)
+
+	return nil
+}
+
+func (a *Autopilot) adjudicateRemoval(s VoterRegistry, voterCountProvider func() int) []raft.ServerID {
+	var ids []raft.ServerID
+	maxRemoval := (voterCountProvider() - 1) / 2
+	minQuorum := a.delegate.AutopilotConfig().MinQuorum
+
+	for id, v := range s {
+		if v != nil && v.IsPotentialVoter() && voterCountProvider()-1 < int(minQuorum) {
+			a.logger.Debug("will not remove server node as it would leave less voters than the minimum number allowed", "id", id, "min", minQuorum)
+		} else if v.IsCurrentVoter() && maxRemoval < 1 {
+			a.logger.Debug("will not remove server node as removal of a majority of voting servers is not safe", "id", id)
+		} else if v != nil && v.IsCurrentVoter() {
 			maxRemoval--
-			voters--
+			delete(s, id)
```

Comment:
Here's one reason why it's desirable to nest the map inside a struct: it's weird when you start treating the VoterRegistry type as a map, and it exposes an implementation detail that you might prefer not to.

Reply:
I've adjusted the implementation, but at the moment I believe we need this to remove the voter from the registry, so that subsequent calls in the for loop to `PotentialVoters` return accurate numbers.

Reply:
Now tracking the overall updated number of voters locally.